apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,427 @@
+            ets:delete(?ETS_TABLE, Data#data.id),
+            Actions = [{reply, From, {error, ConnectStatus}}],
+            {next_state, connecting, UpdatedData, Actions}
+    end.
+
+connecting_health_check(Data) ->
+    case emqx_resource:call_health_check(Data#data.id, Data#data.mod, Data#data.state) of
+        connected ->
+            UpdatedData = Data#data{status = connected, error = undefined},
+            update_resource(Data#data.id, Data#data.group, UpdatedData),
+            {next_state, connected, UpdatedData};
+        {connected, NewResourceState} ->
+            UpdatedData = Data#data{
+                state = NewResourceState, status = connected, error = undefined
+            },
+            update_resource(Data#data.id, Data#data.group, UpdatedData),
+            {next_state, connected, UpdatedData};
+        ConnectStatus ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, ConnectStatus]),
+            UpdatedData = Data#data{error = ConnectStatus},
+            Actions = [{state_timeout, ?SHORT_HEALTHCHECK_INTERVAL, health_check}],
+            {keep_state, UpdatedData, Actions}
+    end.
+
+perform_connected_health_check(Data) ->
+    case emqx_resource:call_health_check(Data#data.id, Data#data.mod, Data#data.state) of
+        connected ->
+            Actions = [{state_timeout, ?HEALTHCHECK_INTERVAL, health_check}],
+            {keep_state_and_data, Actions};
+        {connected, NewResourceState} ->
+            UpdatedData = Data#data{
+                state = NewResourceState, status = connected, error = undefined
+            },
+            update_resource(Data#data.id, Data#data.group, UpdatedData),
+            Actions = [{state_timeout, ?HEALTHCHECK_INTERVAL, health_check}],
+            {keep_state, NewResourceState, Actions};
+        ConnectStatus ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, ConnectStatus]),
+            UpdatedData = Data#data{error = ConnectStatus},
+            ets:delete(?ETS_TABLE, Data#data.id),
+            {next_state, connecting, UpdatedData}
+    end.
+
+data_record_to_external_map_with_metrics(Data) ->
+    #{
+        id => Data#data.id,
+        mod => Data#data.mod,
+        config => Data#data.config,
+        status => Data#data.status,
+        state => Data#data.state,
+        metrics => get_metrics(Data#data.id)
+    }.
+
+make_test_id() ->
+    RandId = iolist_to_binary(emqx_misc:gen_id(16)),
+    <<?TEST_ID_PREFIX, RandId/binary>>.
+
+-spec wait_for_resource_ready(instance_id(), integer()) -> ok | timeout.
+wait_for_resource_ready(InstId, WaitTime) ->
+    do_wait_for_resource_ready(InstId, WaitTime div ?TIME_DIVISOR).
+
+do_wait_for_resource_ready(_InstId, 0) ->
+    timeout;
+do_wait_for_resource_ready(InstId, Retry) ->
+    case ets_lookup(InstId) of
+        {ok, _Group, #{status := connected}} ->
+            ok;
+        _ ->
+            timer:sleep(?WAIT_FOR_RESOURCE_DELAY),
```

Comment:
Small nit: since `?TIME_DIVISOR` must be the same as `?WAIT_FOR_RESOURCE_DELAY` for the total timeout delay to match what was passed in as argument, maybe these two macros could be unified into one?